connectivity/FEATURE_BLE/source/cordio/source/GattServerImpl.cpp

```diff
@@ -1146,8 +1148,23 @@ uint8_t GattServer::atts_read_cb(
             return read_auth_params.authorizationReply & 0xFF;
         }
 
-        pAttr->pValue = read_auth_params.data;
-        *pAttr->pLen = read_auth_params.len;
+        /* if new data provided copy into the attribute value buffer */
+        if (read_auth_params.data) {
+            if (read_auth_params.len > pAttr->maxLen) {
+                tr_error("Read authorisation callback set length larger than maximum attribute length, "
+                         "cannot copy data");
+                err = ATT_ERR_UNLIKELY;
+            } else {
+                memcpy(pAttr->pValue, read_auth_params.data, read_auth_params.len);
```

Comment:
Maybe I'm being dumb but why do we need `memcpy` here? Is a simple assignment not cleaner?

Reply:
You're not dumb but no, a simple assignment will not be cleaner and certainly not correct. In C++ assigning pointer does not copy the memory they point to it just copies the value of the pointer (https://stackoverflow.com/questions/39938648/copy-one-pointer-content-to-another).

Reply:
It was the original bug (one of two).

Reply:
I read that thread and had a aha moment and finally understand the second bug :smile: The problem is once we return from [atts_read_cb](https://github.com/ARMmbed/mbed-os/blob/376fda5bf59c37bfcc0615e6a902c9ac74a96e88/connectivity/FEATURE_BLE/source/cordio/source/GattServerImpl.cpp#L1119), params->data and pAttr->pValue are pointing to the same memory location.

Reply:
@noonfom If the buffer returned to the stack is not statically allocated, the application releases it when `handleDataReadEvent` is called. This happens before the return from `atts_read_cb` meaning the cordio stack hasn't copied yet the data. 

The memcpy fixes a potential use after free.

Reply:
But, [read_auth_params](https://github.com/ARMmbed/mbed-os/blob/376fda5bf59c37bfcc0615e6a902c9ac74a96e88/connectivity/FEATURE_BLE/source/cordio/source/GattServerImpl.cpp#L1129) goes out of scope once we return from atts_read_cb, so actually we create a new pointer on the stack next time `atts_read_cb` gets called...